### PR TITLE
Redesigned pixelatorR classes for improved I/O

### PR DESCRIPTION
**Pull request by ludvigla**
## Description

This PR includes a number of changes to improve I/O performance in pixelatorR. 

The main goal is to re-design the pixelatorR classes and their emthods to enable more efficient reading of edgelist data from pxl files. The new design should mitigate previous issues related to merge/subset/rename operations which created copies of edgelists in a local directory. 

### pixelatorR classes

At the moment, we now have two options to work with, which can be controlled with the global option "Seurat.object.assay.version". This is an option provided by Seurat which allows you to change between using the Assay or Assay5 class for storing count data (see [this](https://satijalab.org/seurat/articles/seurat5_essential_commands) article). The two classes provided in pixelatorR are described in brief below:

The CellGraphAssay will be used if `options(Seurat.object.assay.version = "v3")` and CellGraphAssay5 will be used if `options(Seurat.object.assay.version = "v5")`.  The class objects no longer store paths to edgelist directories. Instead, they store simple tables (`tbl_df`) which maps the current component IDs to a pxl file. Whenever we need to load edgelist data, we can use this table to decide what pxl files to search in. Then we can extract the edgelist.parquet file to a temporary location, read the edgelist data we need, and then remove the edgelist.parquet file. The only difference between CellGraphAssay and CellGraphAssay5 is that they inherit from the Assay or Assay5 Seurat class. 

On top of this, we now also have a virtual class called MPXAssay which is a superclass of CellGraphAssay and CellGraphAssay5. The reason is that the methods for the two classes are nearly identical so instead of having to write a method for both, we now only need to define methods for the MPXAssay class.

NB: we could also stick with a single class which can hold either an Assay or an Assay5 object, but I think it's better to keep these separate for clarity.

### methods

The following methods have been updated:

- merge : merges MPXAssay objects
- subset : filters MPXAssay objects
- RenameCells : renames cell IDs of MPXAssay objects
- LoadCellGraphs : loads component graphs in memory directly from pxl files
- PolarizationScores/PolarizationScores<- : get/set polarization scores in MPXAssay objects
- ColocalizationScores/ColocalizationScores<- : get/set colocalization scores in MPXAssay objects
- PolarizationScoresToAssay : create a new Assay or Assay5 object with polarization scores stored in matrix format
- ColocalizationScoresToAssay : create a new Assay or Assay5 object with colocalization scores stored in matrix format
- RemoveCellGraphs : remove cell graphs from an MPXAssay object
- KeepLargestComponent : filters cell graphs to keep the largest connected component in a CellGraphAssay5 object

### hdf5r

I also took the opportunity to replace the rhdf5 R package from BioConductor with the hdf5r R package from CRAN. CRAN provides MacOS binaries for arm64 (mac chip) which should hopefully facilitate moving away from using rosetta to emulate x86_64.

### Export data to a .pxl file

Added a new function `WriteMPX_pxl_file` which takes a `Seurat` object as input and exports its content to a .pxl file. The steps are detailed below:

- A temporary directory is created to keep the content of the final .pxl file
- Spatial metrics (polarization/colocalization) scores are exported directly from the CellGraphAssay5 object stored in the `Seurat` object to parquet files
- Edge lists are first loaded from the .pxl files found in the fs_map tibble stored in the CellGraphAssay object. Then the component IDs are updated and the edge lists are merged and exported to a new parquet file.
- Meta data files (.json) are loaded and merged and then reexported to a new .json file.
- The raw counts, component meta data and marker meta data are exported from the CellGraphAssay5 to a parquet file

The temporary directory is then zipped and saved to the `file` specified in `WriteMPX_pxl_file`.

### Tests

Some tests are now run twice to check both MPXAssay types. The assay type global option is set at the top of the corresponding test files.

Partially fixes: # EXE-1549

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue).
- [x] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] This change requires a documentation update.

## How Has This Been Tested?

Unit tests have been created for the CellGraphAssay5 methods.

## PR checklist:

- [x] This comment contains a description of changes (with reason).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have checked my code and documentation and corrected any misspellings.
- [x] I have run R CMD check on the package and it passes without errors or warnings (notes can be acceptable if motivated)
- [x] I have documented any significant changes to the code in [CHANGELOG.md](../CHANGELOG.md)

### CODE REVIEW

File: R/load_cell_graphs.R [hunk 1]
```diff
@@ -214,6 +262,142 @@ LoadCellGraphs.CellGraphAssay <- function (
...
+
+    # Split data into chunks determined by chunk_size
+    id_data_chunks <- fs_map_nested_filtered$data[[i]] %>%
+      mutate(group = ceiling(seq_len(n())/chunk_size)) %>%
+      rename(component = current_id) %>%
+      group_by(group) %>%
+      group_split()
+
+    # Read cellgraphs
+    cg_list <- pbapply::pblapply(id_data_chunks, function(id_chunk) {
```
Comment (maxkarlsson) [inline R/load_cell_graphs.R:357]: Does pbapply parallelize successfully across platforms now? It used to have issues with that on Windows.

File: R/load_cell_graphs.R [hunk 2]
```diff
@@ -214,6 +262,142 @@ LoadCellGraphs.CellGraphAssay <- function (
...
+
+      # Filter edgelist data for the current chunk
+      # Note that we use original_id which are the original
+      # MPX component ids. current_id are the ids currenty used
+      # for components in object
+      edgelist_data <- ar %>%
```
Comment (maxkarlsson) [inline R/load_cell_graphs.R:363]: Why is the edgelist read into memory but the graph construction is performed in chunks? Isn't the edgelist itself the item that takes up a lot of RAM? 

The filtering per chunk seems like it could be intense, since the edgelist needs to be parsed once for every chunk (?)

I am not sure about this though, this is just an observation. What do you think?

File: R/load_cell_graphs.R [hunk 3]
```diff
@@ -232,22 +416,24 @@ LoadCellGraphs.Seurat <- function (
...
   chunk_length = 10,
+  cl = NULL,
   verbose = TRUE,
   ...
 ) {
 
   # Use default assay if assay = NULL
   if (!is.null(assay)) {
-    stopifnot("'assay' must be a character of length 1" = is.character(assay) & (length(assay) == 1))
+    stopifnot("'assay' must be a character of length 1" = is.character(assay) && (length(assay) == 1))
```
Comment (maxkarlsson) [inline R/load_cell_graphs.R:426]: What about:
```
# Use default assay if assay = NULL
  assay <- assay %||% DefaultAssay(object)
```

File: R/load_data.R [hunk 4]
```diff
@@ -45,29 +45,32 @@ ReadMPX_counts <- function (
...
     abort(glue("Invalid file format .{.file_ext(filename)}. Expected a .pxl file."))
   }
 
   # Read temporary file
-  anndata_hier <- h5read(filename, "/")
-
-  # Remove temporary file
-  if (endsWith(original_filename, ".pxl")) file.remove(filename)
+  hd5_object <- hdf5r::H5File$new(filename, "r")
+  #anndata_hier <- h5read(filename, "/")
```
Comment (maxkarlsson) [inline R/load_data.R:59]: Should this comment be removed?

File: R/load_data.R [hunk 5]
```diff
@@ -45,29 +45,32 @@ ReadMPX_counts <- function (
...
   # Extract contents
-  X <- anndata_hier$X
-  colnames(X) <- anndata_hier$obs$component
-  rownames(X) <- anndata_hier$var$marker
+  X <- hd5_object[["X"]]$read()
+  colnames(X) <- hd5_object[["obs"]][["component"]]$read()
+  rownames(X) <- hd5_object[["var"]][["marker"]]$read()
+  # X <- anndata_hier$X
+  # colnames(X) <- anndata_hier$obs$component
+  # rownames(X) <- anndata_hier$var$marker
```
Comment (maxkarlsson) [inline R/load_data.R:67]: And these comments?

File: R/load_data.R [hunk 6]
```diff
@@ -45,29 +45,32 @@ ReadMPX_counts <- function (
...
+  X <- X[seq_len(nrow(X)), seq_len(ncol(X))]
```
Comment (maxkarlsson) [inline R/load_data.R:68]: Any reason to use `seq_len(nrow(X))` rather than `1:nrow(X)`?

File: R/load_data.R [hunk 7]
```diff
@@ -188,54 +208,53 @@ ReadMPX_Seurat <- function (
...
+      getOption("Seurat.object.assay.version", "v3"),
+      "v3" = CreateAssayObject(counts = X),
+      "v5" = CreateAssay5Object(counts = X)
+    )
     Key(cg_assay) <- paste0(assay, "_")
   }
 
   # Create Seurat object
-  seur_obj <- CreateSeuratObject(counts = cg_assay, assay = assay, ...)
+  seur_obj <- CreateSeuratObject(counts = cg_assay, assay = assay)#, ...)
```
Comment (maxkarlsson) [inline R/load_data.R:220]: A tailing comment here to remove?

File: R/load_data.R [hunk 8]
```diff
@@ -188,54 +208,53 @@ ReadMPX_Seurat <- function (
...
-      for (name in names(newdata)) {
-        if (length(newdata[[name]]) == dim(anndata_hier$X)[2]) next
-        indicies <- newdata[[name]]$codes + 1
-        categories <- newdata[[name]]$categories
-        newdata[[name]] <- factor(categories[indicies], categories)
+    names(hd5_object[["obs"]]) %>% lapply(function(nm) {
+      if (length(names(hd5_object[["obs"]][[nm]])) == 2) {
+        indices <- hd5_object[["obs"]][[nm]][["codes"]]$read() + 1
+        categories <- hd5_object[["obs"]][[nm]][["categories"]]$read()
+        col <- tibble(!!sym(nm) := factor(categories[indices], categories))
```
Comment (maxkarlsson) [inline R/load_data.R:231]: What is going on here?

File: R/pivot_tables.R [hunk 9]
```diff
@@ -255,32 +258,35 @@ ColocalizationScoresToAssay.CellGraphAssay <- function (
...
-  tofillMat[, colnames(col_scores_wide_format)] <- col_scores_wide_format
+#' @rdname ColocalizationScoresToAssay
+#' @method ColocalizationScoresToAssay CellGraphAssay5
+#'
+#' @export
+#'
+ColocalizationScoresToAssay.CellGraphAssay5 <- function (
+  object,
+  values_from = c("pearson_z", "pearson", "pearson_mean",
+                  "jaccard_mean", "jaccard", "jaccard_z"),
```
Comment (maxkarlsson) [inline R/pivot_tables.R:280]: The pearson_mean and jaccard_mean are the random means estimated from permutations. Should they really be retrievable?

File: R/pivot_tables.R [hunk 10]
```diff
@@ -357,3 +363,50 @@ ColocalizationScoresToAssay.Seurat <- function (
...
+  if (is.null(spatial_metric_table)) {
+    abort(glue("'{metric}' scores are missing from '{class(object)}'"))
+  }
+
+  # Validate input
+  values_from <- match.arg(values_from, choices = switch(
+    metric,
+    "polarization" = c("morans_z", "morans_i"),
+    "colocalization" = c("pearson_z", "pearson", "pearson_mean",
+                         "jaccard_mean", "jaccard", "jaccard_z")
```
Comment (maxkarlsson) [inline R/pivot_tables.R:389]: ping here, regarding the previous comment

File: R/pivot_tables.R [hunk 11]
```diff
@@ -357,3 +363,50 @@ ColocalizationScoresToAssay.Seurat <- function (
...
+
+  # Make sure that the components match
+  # Create an empty matrix (all 0's)
+  tofillMat <- matrix(data = 0,
+                      nrow = nrow(spatial_metric_wide_format),
+                      ncol = ncol(object),
+                      dimnames = list(rownames(spatial_metric_wide_format), colnames(object)))
+
+  # Fill matrix where it overlaps
+  # Any missing columns will be kept as 0's
```
Comment (maxkarlsson) [inline R/pivot_tables.R:408]: Zero feels like a decent default, but in actuality it is a missing value. Should the user be able to control the fill value?

File: R/pivot_tables.R [hunk 12]
```diff
@@ -357,3 +363,50 @@ ColocalizationScoresToAssay.Seurat <- function (
...
+  tofillMat[, colnames(spatial_metric_wide_format)] <- spatial_metric_wide_format
```
Comment (maxkarlsson) [inline R/pivot_tables.R:409]: Why is this necessary when `pivot_wider` is used? It should fill in with zeros anyway?

Comment (maxkarlsson) [reply R/pivot_tables.R:409]: The function `PolarizationScoresToAssay.data.frame` fills with 0's using `pivot_wider` and then again explicitly: 

```
# Cast data.frame to wide format
  pol_scores_wide_format <- object %>%
    pivot_wider(id_cols = "marker",
                names_from = "component",
                values_from = all_of(values_from),
                values_fill = 0) %>%
    data.frame(row.names = 1) %>%
    as.matrix()
  # Replace missing values with 0
  pol_scores_wide_format[is.na(pol_scores_wide_format)] <- 0
```

Perhaps we should remove the explicit replacement up there.

**Review by maxkarlsson — APPROVED** [review]
I think it looks good! I have some questions as usual. Maybe we could meet and discuss, especially since this is quite a chunky PR.

File: R/load_cell_graphs.R [hunk 13]
```diff
@@ -214,6 +262,142 @@ LoadCellGraphs.CellGraphAssay <- function (
...
 }
 
+#' @param force Force load graph(s) if they are already loaded
+#' @param chunk_size An integer specifying a number of components to load
+#' in chunks.
+#' @param cl A cluster object created by makeCluster, or an integer
+#' to indicate number of child-processes (integer values are ignored
+#' on Windows) for parallel evaluations. See Details on performance
+#' in the documentation for \code{pbapply}. The default is NULL,
+#' which means that no parallelization is used.
```
[the remark this answers is not in the document]
Comment (ludvigla) [reply R/load_cell_graphs.R:272]: Good question. I like to document parameters shared by all methods in generics.R but if they are specific for a method, I try to put them in the documentation for that specific method. Just a tiny bit easier to understand where the parameter belongs.

File: R/load_cell_graphs.R [hunk 14]
```diff
@@ -214,6 +262,142 @@ LoadCellGraphs.CellGraphAssay <- function (
...
+      loaded_graphs <- !sapply(slot(object, name = "cellgraphs")[cells], is.null)
+    }
+    cells_to_load <- setdiff(cells, cells[loaded_graphs])
+    if (verbose && check_global_verbosity() & (length(cells_to_load) < length(cells)))
+      cli_alert_info(glue("{length(cells) - length(cells_to_load)} CellGraphs loaded.",
+                          " Loading remaining {length(cells_to_load)} CellGraphs."))
+  }
+
+  # Find where components are located
+  fs_map_unnested_filtered <- slot(object, name = "fs_map") %>%
```
[the remark this answers is not in the document]
Comment (ludvigla) [reply R/load_cell_graphs.R:321]: This is the new slot provided in the CellGraphAssay5 class. It's a tibble with 3 columns:

- id_map : a list of tibbles
   - each tibble has a current_id column with the current component IDs and a original_id column with the original MPX IDs
- sample : an integer for each sample (you get multiple samples here if the object is merged)
- pxl_file : paths to the pxl files

[the remark this answers is not in the document]
Comment (ludvigla) [reply R/load_cell_graphs.R:321]: This is all information we need to know in order to map a set of component IDs to a specific pxl file. When merging data, we only need to merge these tables instead of having to go into the edgelists and fiddle around.

File: R/load_cell_graphs.R (hunk 1 above)
Comment (ludvigla) [reply R/load_cell_graphs.R:357]: I have set the the `cl` parameter to NULL which defaults to serial processing. Users can pass an integer to decide on the number of cores to use which should work on Mac OS but is ignored on Windows. On windows, you can instead create a cluster and pass it with cl. To it's up to the user to decide how to handle parallelization.

File: R/load_cell_graphs.R (hunk 2 above)
Comment (ludvigla) [reply R/load_cell_graphs.R:363]: Yeah it's a bit of a tradeoff I guess. Here, we only read the necessary columns (upia, upib, marker, component) which lowers the memory usage a bit. We are also only reading 1 edgelist at the time, so if you have merged data, we only have to read the edgelist from one sample at the time. One that edgelist has been handled, we free up memory which can be used to load from the next sample. 

Given the size of our edgelists I think this is ok, but I was thinking I could implement an alternative strategy which only fetches chunks from the data stored on disk. It is just much slower to send querys to disk, but could be a good option for massive data sets. When I was playing around with some data sets, the memory usage wasn't too bad, but I didn't try any massive data sets. We could at least have an option for this. Also, we cannot parallelize the conversion from edgelist to graphs if we read directly from disk unless we first split the files somehow. But if we reformat the edgelists to use hive-style partitioning, I think the latter option would be much better!

File: R/load_cell_graphs.R (hunk 3 above)
Comment (ludvigla) [reply R/load_cell_graphs.R:426]: Not sure if I follow :-O It's there after the else statement.

File: R/load_data.R (hunk 8 above)
Comment (ludvigla) [reply R/load_data.R:231]: Hehe I'll try to add some comments in the code.

I have replaced rhdf5 with hdf5r here. It has a pretty strange API but I think it will be useful for us to get rid of rhdf5 since it doesn't play well with arm chip sets. 

But what the code does is to go through the "obs" table in the anndata file and convert them into tibble columns. This part:

````
indices <- hd5_object[["obs"]][[nm]][["codes"]]$read() + 1
        categories <- hd5_object[["obs"]][[nm]][["categories"]]$read()
````

converts dictionaries into factors. In the anndata object, the dictionaries have a codes vector (values) and a categories vector (keys).  If an element in "obs" has length 2, it's a dictionary that should be converted into a factor.

This stuff `!!sym(nm) := ` is just a trick to create a name from a variable. For example, if `nm = "leiden"`, these two solutions should be equivalent:

````
tibble(leiden = factor(categories[indices], categories))
tibble(!!sym(nm) := = factor(categories[indices], categories))
````

File: R/pivot_tables.R (hunk 9 above)
Comment (ludvigla) [reply R/pivot_tables.R:280]: I think that's a fair point. I would much rather trim down this list! Maybe "pearson_z" and "pearson" is enough?

File: R/pivot_tables.R (hunk 11 above)
Comment (ludvigla) [reply R/pivot_tables.R:408]: The problem is that (unless this has been changed in Seurat 5) Assays do not allow missing values. The idea with these pivot functions was to create assays that can be stored in Seurat objects, so we have to set these values to 0. This was one of the concerns I had with this conversion, but I still think that it is quite useful to be able to store the spatial metrics in assays. It would get even worse if we decide to filter spatial metrics based on marker counts. Not sure what to do about it though :-/

File: R/pivot_tables.R (hunk 12 above)
Comment (ludvigla) [reply R/pivot_tables.R:409]: Right, must have forgotten to remove that!

File: R/load_data.R (hunk 8 above)
Comment (maxkarlsson) [reply R/load_data.R:231]: Check, thanks!

File: R/pivot_tables.R (hunk 9 above)
Comment (maxkarlsson) [reply R/pivot_tables.R:280]: I would think so, yes!

File: R/pivot_tables.R (hunk 11 above)
Comment (maxkarlsson) [reply R/pivot_tables.R:408]: Okay, that's a fair point! But for example the `counts` slot is a sparse matrix? But the missing values are interpreted as zeros. 

I suppose the issue arises when e.g. PCA is applied on the polarization matrix. One would need to use a sparse versions of algorithms and interpret zeros as missing values to avoid bias from zeros. It seems quite farfetched at the current time.

File: R/load_data.R (hunk 6 above)
Comment (ludvigla) [reply R/load_data.R:68]: It's generally [adviced](https://win-vector.com/2018/02/19/r-tip-use-seq_len-to-avoid-the-backwards-sequence-bug/#:~:text=Use%20seq_len()%20to%20avoid%20the%20backwards%20sequence%20trap.&text=Notice%20the%20above%20example%20built,x)%20%E2%80%9D%20is%20often%20wrong.) to use seq_len and seq_along instead of 1:X. In this case it shouldn't matter, but I have tried to be consistent with it :-)

File: R/load_data.R (hunk 7 above)
Comment (ludvigla) [reply R/load_data.R:220]: Ah thanks!

File: R/pivot_tables.R (hunk 11 above)
Comment (ludvigla) [reply R/pivot_tables.R:408]: Yeah I think you're right that certain methods might be unfit to run on the spatial metrics as they are represented now. Missing values is a bit of a pin :-/

Comment (maxkarlsson) [reply R/pivot_tables.R:408]: Yeah, but we'll get to that in time! It is almost a philosophical question how to represent the position of something that doesn't exist, haha. Maybe it is zero...?

File: R/load_data.R (hunk 6 above)
Comment (maxkarlsson) [reply R/load_data.R:68]: Ok, check!

File: R/load_cell_graphs.R (hunk 3 above)
Comment (maxkarlsson) [reply R/load_cell_graphs.R:426]: Ok, nevermind!

File: tests/testthat/Rplots.pdf
Comment (maxkarlsson) [inline tests/testthat/Rplots.pdf:1]: Should this file be commited?

**Review by maxkarlsson — APPROVED** [review]
Looks good to me! I will continuously test this going forward. Great work making these large and important changes. It will be a great improvement!

Only one comment to check whether the pdf should be commited.

File: R/write_pxl_file.R [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+      is(cg_assay, "MPXAssay")
+  )
+
+  # fetch and validate fs_map
+  fs_map <- cg_assay@fs_map
+  .validate_fs_map(fs_map)
+
+  # Delete the file only if it's not in use
+  # by the current object
+  if (fs::file_exists(file)) {
```
Comment (maxkarlsson) [inline R/write_pxl_file.R:107]: If the file exists, 
check if file is in the file system map
If it is, cancel the file creation
Otherwise, remove the file

Why is the fiile removed? Is it because it should be overwritten and replaced?

File: R/write_pxl_file.R [hunk 16]
```diff
@@ -0,0 +1,499 @@
...
+    if (file %in% fs_map$pxl_file) {
+      abort(glue(
+        "The selected file name '{col_br_blue(file)}' is currently in use ",
+        "by the input object. You need to select a different file name."
+      ))
+    }
+    fs::file_delete(file)
```
Comment (maxkarlsson) [inline R/write_pxl_file.R:114]: A general comment; why is `fs` used instead of base functions?

File: R/write_pxl_file.R [hunk 17]
```diff
@@ -0,0 +1,499 @@
...
+  cli_rule(left = "Creating .pxl file")
+
+  # Export spatial metrics
+  .write_spatial_metrics_to_parquet(object = object, pxl_folder = pxl_folder)
+
+  # Update and write edgelist data
+  .merge_and_write_edgelists_to_parquet(fs_map = fs_map, pxl_folder = pxl_folder)
+
+  # Extract and merge meta data
+  .write_json_meta_data(fs_map = fs_map, pxl_folder = pxl_folder)
```
Comment (maxkarlsson) [inline R/write_pxl_file.R:130]: Looking at this function, it looks like the json file is generated from a list of strings from the original json files, if there are multiple. Is that correct? How does that work with `pixelator`?

**Review by maxkarlsson — APPROVED** [review]

File: R/write_pxl_file.R (hunk 15 above)
Comment (ludvigla) [reply R/write_pxl_file.R:107]: Yeah you're right. It doesn't have to be deleted to proceed. I added it as an extra precaution in case the file is write-protected.

File: R/write_pxl_file.R (hunk 16 above)
Comment (ludvigla) [reply R/write_pxl_file.R:114]: We could switch to base R, but I prefer fs because it is IMO more rigorous. For example:

"Compared to [file.remove](http://127.0.0.1:10927/help/library/fs/help/file.remove) they always fail if they cannot delete the object rather than changing return value or signalling a warning."

We could of course handle warnings manually, but I think this is a nice feature of fs that their function always errors if the command fails.

File: R/write_pxl_file.R (hunk 17 above)
Comment (ludvigla) [reply R/write_pxl_file.R:130]: The metadata.json files are parsed in R with jsonlite which returns the content as a list object. I've tried to mimic the `pixelator` behaviour by adding a top level field called "sample". When the merged list is exported with jsonlite, the list structure is translated in to json format. This is what it could look like after merging:

````
{
    "samples": {
        "sample1": {
            "version": "0.12.0",
            "sample": "Sample07_SG_CD3_S7_001",
            "file_format_version": 1,
            "analysis": {
                "params": {
                    "compute_polarization": true,
                    "compute_colocalization": true,
                    "use_full_bipartite": false,
                    "polarization_normalization": "clr",
                    "polarization_binarization": false,
                    "colocalization_transformation": "log1p",
                    "colocalization_neighbourhood_size": 1,
                    "colocalization_n_permutations": 50,
                    "colocalization_min_region_count": 5
                }
            }
        },
        "sample2": {
            "version": "0.12.0",
            "sample": "Sample02_SG_unstimulated_S2_001",
            "file_format_version": 1,
            "analysis": {
                "params": {
                    "compute_polarization": true,
                    "compute_colocalization": true,
                    "use_full_bipartite": false,
                    "polarization_normalization": "clr",
                    "polarization_binarization": false,
                    "colocalization_transformation": "log1p",
                    "colocalization_neighbourhood_size": 1,
                    "colocalization_n_permutations": 50,
                    "colocalization_min_region_count": 5
                }
            }
        },
        "sample3": {
            "version": "0.12.0",
            "sample": "Uropod_CD54_Rantes_coated_S3_001",
            "file_format_version": 1,
            "analysis": {
                "params": {
                    "compute_polarization": true,
                    "compute_colocalization": true,
                    "use_full_bipartite": false,
                    "polarization_normalization": "clr",
                    "polarization_binarization": false,
                    "colocalization_transformation": "log1p",
                    "colocalization_neighbourhood_size": 1,
                    "colocalization_n_permutations": 50,
                    "colocalization_min_region_count": 5
                }
            }
        }
    }
}
````

But thanks for pointing this out. I need to double check that the output look as expected.

File: R/write_pxl_file.R (hunk 16 above)
Comment (maxkarlsson) [reply R/write_pxl_file.R:114]: Ok, that sounds like a good motivation to me!

File: R/cell_count_plot.R
Comment (vincent-van-hoef) [inline R/cell_count_plot.R:1]: Neat, I somehow missed this function existed :)

File: R/load_data.R [hunk 18]
```diff
@@ -113,14 +107,12 @@ ReadMPX_counts <- function (
...
 #' library(pixelatorR)
-#' # Set arrow data output directory to temp for tests
-#' options(pixelatorR.arrow_outdir = tempdir())
 #'
 #' # Load example data as a Seurat object
 #' pxl_file <- system.file("extdata/five_cells",
 #'                         "five_cells.pxl",
 #'                         package = "pixelatorR")
-#' seur_obj <- ReadMPX_Seurat(pxl_file, overwrite = TRUE)
+#' seur_obj <- ReadMPX_Seurat(pxl_file)
```
Comment (vincent-van-hoef) [inline R/load_data.R:115]: I get a warning `Warning: Data is of class matrix. Coercing to dgCMatrix.` for every call to ReadMPX_Seurat. No big deal, but perhaps people will get worried they are doing something wrong?

**Review by vincent-van-hoef — APPROVED** [review]
Great work! Simplified things a lot. I have been playing around with it a bit on the new app note data and so far it works nicely. I'll inform you if that changes :)

File: R/write_pxl_file.R (hunk 15 above)
Comment (maxkarlsson) [reply R/write_pxl_file.R:107]: precautions are good! Let's keep it

File: R/load_data.R (hunk 18 above)
Comment (ludvigla) [reply R/load_data.R:115]: Yeah I have noticed that too. I can add a conversion step to make sure that the matrix is always of class `dgCMatrix` before running `CreateAssayObject`.

File: R/write_pxl_file.R (hunk 16 above)
Comment (ludvigla) [reply R/write_pxl_file.R:114]: Ok then I'll keep it as is!

File: R/write_pxl_file.R (hunk 17 above)
Comment (ludvigla) [reply R/write_pxl_file.R:130]: Ok so there were some issue with the merging which should be resolved now! "samples" was called "sample" and I added `auto_unbox` to unbox all atomic vectors of length 1. Without this, all vectors in the json file are formatted as arrrays. For instance, without it, it would look something like this:

````
"colocalization_n_permutations": [50]
````

instead of:
````
"colocalization_n_permutations": 50
````

I don't think it should matter, but adding this makes the file more readable.
